activesupport/lib/active_support/callbacks.rb

```diff
@@ -318,33 +283,6 @@ def self.halting(callback_sequence, user_callback)
           end
         end
         private_class_method :halting
-
-        def self.conditional(callback_sequence, user_callback, user_conditions)
-          callback_sequence.around do |env, &run|
-            target = env.target
-            value  = env.value
-
-            if user_conditions.all? { |c| c.call(target, value) }
-              user_callback.call(target, value) {
-                run.call.value
-              }
-              env
-            else
-              run.call
-            end
-          end
-        end
-        private_class_method :conditional
-
-        def self.simple(callback_sequence, user_callback)
-          callback_sequence.around do |env, &run|
-            user_callback.call(env.target, env.value) {
-              run.call.value
-            }
-            env
-          end
-        end
-        private_class_method :simple
```

Comment:
All the code above fixes #21122 and it's a slightly different problem than what this PR is addressing.

For this reason, I have made a separate PR #21616 which only deals with removing the unused code above.

If #21616 gets merged into master, then you can rebase this PR on top of that and these deletions should be gone.